ml-agents-envs/mlagents/envs/base_env.py

```diff
@@ -0,0 +1,305 @@
+    simulation step.
+     - obs is a list of numpy arrays observations collected by the group of
+     agent.
+     - reward is a float. Corresponds to the rewards collected by the agent
+     since the last simulation step.
+     - done is a bool. Is true if the Agent was terminated during the last
+     simulation step.
+     - max_step is a bool. Is true if the Agent reached its maximum number of
+     steps during the last simulation step.
+     - agent_id is an int and an unique identifier for the corresponding Agent.
+     - action_mask is an optional list of one dimensional array of booleans.
+     Only available in multi-discrete action space type.
+     Each array corresponds to an action branch. Each array contains a mask
+     for each action of the branch. If true, the action is not available for
+     the agent during this simulation step.
+    """
+
+    obs: List[np.array]
+    reward: float
+    done: bool
+    max_step: bool
+    agent_id: int
+    action_mask: Optional[List[np.array]]
+
+
+class BatchedStepResult:
+    """
+    Contains the data a group of similar Agents collected since the last
+    simulation step. Note that all Agents do not necessarily have new
+    information to send at each simulation step. Therefore, the ordering of
+    agents and the batch size of the BatchedStepResult are not fixed across
+    simulation steps.
+     - obs is a list of numpy arrays observations collected by the group of
+     agent. Each obs has one extra dimension compared to StepResult: the first
+     dimension of the array corresponds to the batch size of
+     the group.
+     - reward is a float vector of length batch size. Corresponds to the
+     rewards collected by each agent since the last simulation step.
+     - done is an array of booleans of length batch size. Is true if the
+     associated Agent was terminated during the last simulation step.
+     - max_step is an array of booleans of length batch size. Is true if the
+     associated Agent reached its maximum number of steps during the last
+     simulation step.
+     - agent_id is an int vector of length batch size containing unique
+     identifier for the corresponding Agent. This is used to track Agents
+     across simulation steps.
+     - action_mask is an optional list of two dimensional array of booleans.
+     Only available in multi-discrete action space type.
+     Each array corresponds to an action branch. The first dimension of each
+     array is the batch size and the second contains a mask for each action of
+     the branch. If true, the action is not available for the agent during
+     this simulation step.
+    """
+
+    def __init__(self, obs, reward, done, max_step, agent_id, action_mask):
+        self.obs: List[np.array] = obs
+        self.reward: np.array = reward
+        self.done: np.array = done
+        self.max_step: np.array = max_step
+        self.agent_id: np.array = agent_id
```

Comment:
Since agent_id is a string, when I print the values of the agent_id it's weird (lots of negative values). Should we make this a List[str]?

Reply:
Object.GetInstanceId() returns an int `it's range is -2,147,483,648 to 2,147,483,647` according to [this page](https://answers.unity.com/questions/58507/getinstanceid-range-of-possible-values.html). Is it causing problems ?

Reply:
No, just curious - was worried that there was some funky type conversion going on. But it seems fine.